pyro/nn/module.py

```diff
@@ -187,14 +201,14 @@ def __setattr__(self, name, value):
             _pyro_params = self.__dict__['_pyro_params']
             constrained_value, constraint, event_dim = value
             _pyro_params[name] = constraint, event_dim
-            with torch.no_grad():
-                constrained_value = constrained_value.detach()
-                unconstrained_value = transform_to(constraint).inv(constrained_value)
-                unconstrained_value = unconstrained_value.contiguous()
-            unconstrained_value = torch.nn.Parameter(unconstrained_value)
-            unconstrained_value._pyro_event_dim = event_dim
-            unconstrained_name = name + "_unconstrained"
-            setattr(self, unconstrained_name, unconstrained_value)
+            fullname = _make_name(self._pyro_name, name)
+            pyro.param(fullname, constrained_value, constraint=constraint, event_dim=event_dim)
+            unconstrained_param = _PYRO_PARAM_STORE._params[fullname]
+            if not isinstance(unconstrained_param, torch.nn.Parameter):
+                unconstrained_param = torch.nn.Parameter(unconstrained_param)
+                _PYRO_PARAM_STORE._params[fullname] = unconstrained_param
```

Comment:
@fehiepsi the following two lines should ensure `.requires_grad_(...)` on a module attribute is also respected by the param store.

Reply:
Thank Fritz! I'll merge this PR first to see if I need to do something else.

Reply:
@fritzo There is some leakage in this logic:
```
import torch
import pyro

m = pyro.nn.PyroModule()
m.a = pyro.nn.PyroParam(torch.tensor(1.))
m.a = pyro.nn.PyroParam(torch.tensor(0.1))
print(m.a)
```
should return 0.1 instead of 1.

Reply:
Another leakage is
```import torch
import pyro

class Child(pyro.nn.PyroModule):
    def __init__(self, a):
        super().__init__()
        self.a = pyro.nn.PyroParam(a, torch.distributions.constraints.positive)

class Family(pyro.nn.PyroModule):
    def __init__(self):
        super().__init__()
        self.child1 = Child(torch.tensor(1.))
        self.child2 = Child(torch.tensor(2.))

f = Family()
print(f.child2.a)
```
which returns 1 instead of 2

Reply:
Thanks, I'll add those as unit tests.

Reply:
OK, those tests now pass.

Reply:
Thanks! With your latest changes, my deep GP model also works as expected.